File: 63. Unique Paths II/unique-paths-ii.ts
// 63. Unique Paths II (2/22/54352)
// Runtime: 115 ms (25.62%) Memory: 45.34 MB (12.40%) 

function uniquePathsWithObstacles(obstacles: number[][]): number {
    const rows = obstacles.length;
    const cols = obstacles[0].length;
    const counts: number[][] = Array
        .from(
            { length: rows }, 
            () => Array.from({ length: cols }, () => 0)
        );
    
    counts[0][0] = obstacles[0][0] === 0 ? 1 : 0;
    for (let y = 0; y < rows; y++) {
        for (let x = 0; x < cols; x++) {
            if (counts[y][x] === 0 && obstacles[y][x] === 0) {
                counts[y][x] = (y > 0 ? counts[y - 1][x] : 0) + (x > 0 ? counts[y][x - 1] : 0);
            }
        }
    }
    
    return counts[rows - 1][cols - 1];
};

File: 63. Unique Paths II/unique-paths-ii.c
// 63. Unique Paths II (3/1/54352)
// Runtime: 8 ms (19.08%) Memory: 6.12 MB (37.40%) 

int uniquePathsWithObstacles (int** obstacles, int rows, int* cols) {
    uint32_t* counts = calloc(rows * *cols, sizeof(uint32_t));
    
    counts[0] = !obstacles[0][0];
    for (size_t y = 0; y < rows; y++) {
        for (size_t x = 0; x < *cols; x++) {
            if (!(x == 0 && y == 0) && !obstacles[y][x]) {
                counts[y * *cols + x] = (x > 0 ? counts[y * *cols + x - 1] : 0) + (y > 0 ? counts[(y - 1) * *cols + x] : 0);
            }
        }
    }
    
    return counts[rows * *cols - 1];
}
